[FLINK-21355] Send changes to the state changelog

rkhachatryan opened the pull request:
## What is the purpose of the change

Integrate ChangelogBackend with ChangelogWriter.

## Verifying this change

Added unit tests for map and list states.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -395,8 +397,8 @@ private static StateBackend loadChangelogStateBackend(
-                            .getConstructor(StateBackend.class);
-            return constructor.newInstance(backend);
+                            .getConstructor(StateBackend.class, StateChangelogWriterFactory.class);
+            return constructor.newInstance(backend, new InMemoryStateChangelogWriterFactory());
```
I've opened FLINK-21804 to load via plugin manager
and propose to use simple constructor + if/else here for now.

rkhachatryan replied:
Removed this change (as per comment [below](https://github.com/apache/flink/pull/15200#discussion_r640955392))

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/InternalReadOnlyKeyContext.java:
```diff
@@ -0,0 +1,43 @@
+ * key-grouping related information.
+ */
+@Internal
+public interface InternalReadOnlyKeyContext<K> {
```
Why we have to introduce `InternalReadOnlyKeyContext`. Is it possible to just use `InternalKeyContext` instead?


We can use the InternalKeyContext maintained in `AbstractKeyedStateBackend` to initiate `StateChangeLogger`, so that we do not need to have AbstractKeyedStateBackend implement InternalReadOnlyKeyContext?

rkhachatryan replied:
Removed `InternalReadOnlyKeyContext`.

I wanted to prevent unnecessary access to changing the current key. But later during recovery implementation I realized I have to pass `InternalKeyContext` anyways.

curcur commented:
I've done a pass. I've found `ChangelogAggregatingState` has not been integrated with changelog yet? 

I will do another pass on state MERGE_NS/CHANGE_ELEMENT/REMOVE_ELEMENT

after the PR is ready.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java:
```diff
@@ -38,8 +40,23 @@
+
+    ChangelogValueState(
+            InternalValueState<K, N, V> delegatedState, StateChangeLogger<V, N> changeLogger) {
+        super(delegatedState, changeLogger);
```
Good idea.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java:
```diff
@@ -49,17 +66,25 @@ public V value() throws IOException {
 
     @Override
     public void update(V value) throws IOException {
+        changeLogger.stateUpdated(value, currentNamespace);
         delegatedState.update(value);
```
As disccussed offline, write efficiency is a changelog implementation concern.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLoggerImpl.java:
```diff
@@ -0,0 +1,148 @@
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e);
+        }
+    }
```
This is because `clear()` method on State classes doesn't declare any checked exceptions. Declaring it here would increase this boilerplate code (though I agree that API here is non-uniform in that respect).

rkhachatryan commented:
Thanks for the review @curcur, I've rebased the PR and added:
- aggregating state
- PQ state
- logging metadata and other fixes I've made during recovery implementation
Please take a look.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractChangelogState.java:
```diff
@@ -34,10 +34,13 @@
         implements InternalKvState<K, N, V> {
     protected final S delegatedState;
+    protected N currentNamespace;
+    protected KvStateChangeLogger<V, N> changeLogger;
```
`final` ?

rkhachatryan replied:
Good point!

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractChangelogState.java:
```diff
@@ -34,10 +34,13 @@
 abstract class AbstractChangelogState<K, N, V, S extends InternalKvState<K, N, V>>
         implements InternalKvState<K, N, V> {
     protected final S delegatedState;
+    protected N currentNamespace;
```
`currentNamespace`  is not initialized, so it is possible to be null? 

From the context, it seems namespace can not be null, so it is better to check its non-null when using it?
`checkNotNull(namespace, "Namespace");` in `AbstractKeyedStateBackend#getPartitionedState`

rkhachatryan replied:
Yes, you are right.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractChangelogState.java:
```diff
@@ -34,10 +34,13 @@
+    protected KvStateChangeLogger<V, N> changeLogger;
 
-    AbstractChangelogState(S state) {
+    AbstractChangelogState(S state, KvStateChangeLogger<V, N> changeLogger) {
```
@Nonnull for KvStateChangeLogger?

rkhachatryan replied:
I slightly prefer to mark nullable fields as such and assume everythin else non-nullable.
But the `notNull` check is missing. Do you think adding it is enough?

curcur replied:
> I slightly prefer to mark nullable fields as such and assume everythin else non-nullable.
> But `Preconditions.notNull` check is missing. Do you think adding it is enough?

yes, that would also be fine.

curcur reviewed (commented):
I've started to review this PR, but found a lot of stuff related to recovery.

Is it possible to separate the metadata and other fixes related to recovery implementation to a separate PR?

It is difficult to review without the context of "recovery implementation". As we discussed, the recovery part may not be able to get in before 1.13, and I cannot have parts that related to the "recovery part" get in without the "recovery part" is finalized.
Having the "recovery part" would slow down the review

For example, I do not know why metadata is needed for forwarding states.
I do not know why metadata to log is set to true and never set back to false
I think these parts are not-related to this PR

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogWriterFactory.java:
```diff
@@ -30,7 +29,7 @@
         extends AutoCloseable {
 
-    StateChangelogWriter<Handle> createWriter(OperatorID operatorID, KeyGroupRange keyGroupRange);
+    StateChangelogWriter<Handle> createWriter(String operatorID, KeyGroupRange keyGroupRange);
```
I'd prefer `OperatorID` type, but it comes in form of `String` here from `StateBackend.createKeyedStateBackend` which is `PublicEvolving`.
I think changing this `PublicEvolving` API would be overkill.

rkhachatryan commented:
I've split the last commit into two: for logging changes and for logging metadata changes as discussed offline.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/KvStateChangeLogger.java:
```diff
@@ -0,0 +1,26 @@
+
+interface KvStateChangeLogger<State, Namespace> extends StateChangeLogger<State, Namespace> {
+
+    void stateMerged(Namespace target, Collection<Namespace> sources) throws IOException;
```
-> `mergeNamespaces` ?

rkhachatryan replied:
I think `stateMerged` is more consistent with parent interface methods (like `stateUpdated`) and better reflects that namespaces aren't actually merged here, but only logged.

curcur replied:
Things to sync up offline: I am confused why logging "name space merge" should be named as "state Merge"

rkhachatryan replied:
Renamed to `namespacesMerged`.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +94,67 @@ public boolean contains(UK key) throws Exception {
+                            }
+
+                            @Override
+                            public void remove() {
+                                iterator.remove();
+                            }
```
would it be possible to remove from the entries here? 

should we have delegated stateElementRemoved here as well?

rkhachatryan replied:
I don't see why it's impossible. `MapState.entries` doesn't says anything forbidding this.

The `iterator` was already obtained from `delegatedState`, so `remove()` call should be enough.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +94,67 @@ public boolean contains(UK key) throws Exception {
 
     @Override
     public Iterable<Map.Entry<UK, UV>> entries() throws Exception {
-        return delegatedState.entries();
+        Iterator<Map.Entry<UK, UV>> iterator = delegatedState.entries().iterator();
+        final N currentNamespace = getCurrentNamespace();
+        return () ->
+                loggingIterator(
+                        new Iterator<Map.Entry<UK, UV>>() {
+                            @Override
+                            public Map.Entry<UK, UV> next() {
+                                return loggingMapEntry(
+                                        iterator.next(),
+                                        changeLogger,
+                                        changeWriter,
+                                        currentNamespace);
+                            }
+
+                            @Override
+                            public boolean hasNext() {
+                                return iterator.hasNext();
+                            }
+
+                            @Override
+                            public void remove() {
+                                iterator.remove();
+                            }
+                        },
+                        changeLogger,
+                        changeWriter,
+                        currentNamespace);
     }
 
     @Override
     public Iterable<UK> keys() throws Exception {
-        return delegatedState.keys();
+        return loggingIterable(
+                delegatedState.keys(), changeLogger, this::serializeKey, getCurrentNamespace());
```
Is it possible to remove a key without removing its value for a map?
Ask because the loggingIterable is mainly for remove change

rkhachatryan replied:
No, it's impossible. I'm not sure I understand your concern though.

What can happen is:
1. user calls `map.put(1, 1)`
2. `delegatedMap` creates a new `HashMap` for the current stream key, stores it, and calls `put(1, 1)` on it
3. user calls `changeLogMap.keys()`
4. `delegatedMap` returns `hashMap.keySet`, wrapped for logging `ChangelogMapState`
5. user calls `keys.iterator.next().remove(x)` - this is supported by jdk `HashMap` and should be logged. Both key and value are removed.

curcur replied:
Let me rephrase my question:

1. Some backend supports iterator change/remove; but some does not, but this is not the focus, I am asking in the case where iterator change/remove are supported (rocksdb for example)

2. In the case where iterator is supported
Let's say I get an iterator list of keys, and I get an iterator list of values, I remove some keys in the key iterator, and some values in the value iterator

Is it guaranteed that such remove is always paired? (Key, Value)

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +94,67 @@ public boolean contains(UK key) throws Exception {
+                    }
+
+                    @Override
+                    public void remove() {
+                        iterator.remove();
+                    }
```
same question as entries() since it is using entries iterator.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +94,67 @@ public boolean contains(UK key) throws Exception {
 
     @Override
     public Iterator<Map.Entry<UK, UV>> iterator() throws Exception {
-        return delegatedState.iterator();
+        return loggingIterator(
+                delegatedState.iterator(), changeLogger, changeWriter, getCurrentNamespace());
     }
```
would it possible to change the value (not add a value) here?

rkhachatryan replied:
Good point, it can be changed via `MapEntry.setValue`.
I'll reuse the code from `entries()`.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogAggregatingState.java:
```diff
@@ -67,18 +71,20 @@ public OUT get() throws Exception {
     @Override
     public void add(IN value) throws Exception {
         delegatedState.add(value);
+        changeLogger.stateUpdated(delegatedState.getInternal(), getCurrentNamespace());
```
That's interesting. Using `update` to simulate `add` could work, but I guess this is something different than what we want?

It is possible the case that "add" could be small, but "update" is huge.

curcur replied:
I guess here we have to have both IN serializer and ACC serializer.

rkhachatryan replied:
We don't know how the accumulator was affected by the added value, so we have to log it (using `logger.stateUpdated(delegated.getInternal())`).
And as we log the accumulator state, there is no need to log the added value.

Accumulator can be either bigger or smaller than the added value (e.g. list and counter). But we don't know it and can't log only the added value (above reason).

rkhachatryan replied:
I've added a clarifying comment as discussed offline.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PqStateChangeLogger.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.io.IOException;
+
+interface PqStateChangeLogger<T> extends StateChangeLogger<T, Void> {
```
If I guess correctly, Pq stands for priority queue, would it be fine to make it specific? It is difficult to guess what Pq stands for unless being familiar with the implmentaion.

rkhachatryan replied:
Yes, PQ=priority queue (and I wanted it to be symmetric to the "kv-" counterpart). 
I'll expand PQ as it creates confusion.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyGroupedPriorityQueue.java:
```diff
@@ -18,24 +18,36 @@
  */
 public class ChangelogKeyGroupedPriorityQueue<T> implements KeyGroupedInternalPriorityQueue<T> {
     private final KeyGroupedInternalPriorityQueue<T> delegatedPriorityQueue;
+    private final PqStateChangeLogger<T> logger;
```
Are these updates writing to the same log? I mean they are using different writers, but they might share the same log write?

If the scope is not specified, how we can differentiate this PQ from the normal change log?

I understand with normal changes can be differentiated using key context, but how is this one?

rkhachatryan replied:
State type is written as part of the metadata (and for each state change, state name is written, so we can relate it to the metadata).

The current key group is written in `AbstractStateChangeLogger.log`, so we can differentiate between the groups on rescaling.

curcur commented:
Main issues:

1. Loader seems a bit too hacky, I am not sure whether we can use it this way, need @pnowojski to take a look
2. Map iterators, have some uncertainties, commented in place
3. Aggregating State, whether we want to use `update` to simulate `add`. I guess this is not our original intention.
4. How to differentiate PriorityQueue changes, also commented in place.

Tests are not reviewed yet, I will review tests now.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java:
```diff
@@ -0,0 +1,122 @@
+
+    void stateAdded(State addedState, Namespace ns) throws IOException;
+
+    void stateCleared(Namespace ns);
```
Yes, that's because `State` doesn't declare checked exceptions.
But after reading your comment I realized that declaring it in `StateChangeLogger.stateCleared` and catching in states makes a more consistent API (at the cost of some code duplication).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/KvStateChangeLoggerImpl.java:
```diff
@@ -0,0 +1,77 @@
+    KvStateChangeLoggerImpl(
+            TypeSerializer<Key> keySerializer,
+            TypeSerializer<Ns> namespaceSerializer,
+            TypeSerializer<State> valueSerializer,
```
I agree, but using `State` would be inconsistent with the parent class `AbstractStateChangeLogger`.
It has e.g. `stateAdded(State addedState)` and `stateElementChanged()`, and using `Value` instead of `State` would be confusing there IMO.
WDYT?

curcur replied:
"AbstractStateChangeLogger<Key, State, Ns>" is confusing as well.

Here the "State" not seems to be a state as well. In its two implementations of 

```
protected abstract void serializeState(State state, DataOutputViewStreamWrapper out)
            throws IOException;
```

One is a `valueSerializer` and the other is PtirotyQueue (time).

Instead of calling it State, they are more like Changelog or Changelog Value instead;

What do you think?

I would appreciate the concept of "State" to be consistent. That's very important to keep the code readable and maintable.

rkhachatryan replied:
I think you're right. 
I've renamed `State` to `Value`, please take a look at b715300b9103f89a54bff16f05bc6e39a7adc085.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -0,0 +1,154 @@
+        /** Scope: key + namespace + element (e.g. user map remove or iterator remove). */
+        REMOVE_ELEMENT((byte) 5),
+        /** Scope: key + namespace, last element. */
+        POLL_ELEMENT((byte) 6);
```
Sure.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -0,0 +1,154 @@
+            this.code = code;
+        }
+
+        static Map<Byte, KvStateChangeLoggerImpl.StateChangeOperation> byCodes =
```
Good point.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -391,8 +393,8 @@ private static StateBackend loadChangelogStateBackend(
-                            .getConstructor(StateBackend.class);
-            return constructor.newInstance(backend);
+                            .getConstructor(StateBackend.class, StateChangelogWriterFactory.class);
+            return constructor.newInstance(backend, new InMemoryStateChangelogWriterFactory());
```
Yes, it is hacky. We discussed it previously and [planned](https://github.com/apache/flink/pull/15200#discussion_r594669021) to implement separately in FLINK-21804 because of the time constraints.

However, I realized that another solution without modifying `StateBackendLoader` would better (passing the writer factory to backend.createKeyedBackend methods). So I'll remove this change alltogether.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.Map;
+
+interface StateChangeLogger<State, Namespace> {
+    static <Namespace, State, StateElement> Iterator<StateElement> loggingIterator(
```
Let me answer in the reverse order:
3. Yes, those methods are needed because of `Iterator.remove` (and `Map.Entry.setValue`). These actions should be logged.
2. `MapState` **does** use these methods, but it also uses `StateChangeLogger` to log other state modifications
1. So to me, these methods are closely related to the interface functionality

However, I'm also fine with extracting it into something like `StateChangeLoggingUtils` if that sounds good to you.

rkhachatryan replied:
I've moved  `loggingMapEntry` to `ChangelogMapState` as discussed offline, as well as `loggingIterable` (as it's only used in map state currently and can easily be extracted);
I've also replaced `loggingIterator` with separate `StateChangeLoggingIterator` class.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogListState.java:
```diff
@@ -39,53 +42,79 @@
 
     @Override
     public void add(V value) throws Exception {
+        if (getValueSerializer() instanceof ListSerializer) {
```
Good point. It would be handled during the deserialization, but according to the contract, null (or empty list in other places) don't change the state.
I'll add the checks.

rkhachatryan commented:
Thanks for the review @curcur,

I've extracted FLINK-21804 to load the writer and  and FLINK-22808 to log metadata. 
I've also addressed your comments, PTAL.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogReducingState.java:
```diff
@@ -65,16 +72,24 @@ public V get() throws Exception {
     @Override
     public void add(V value) throws Exception {
         delegatedState.add(value);
+        changeLogger.valueUpdated(delegatedState.get(), getCurrentNamespace());
```
we should first update changelog and then adding to the delegated state.

rkhachatryan replied:
As discussed offline:
1. in general the order does not matter
2. in this case, it does matter: an updated state value is logged
3. so delegated state should be updated first and changelog - second; this should be done in all places for consistency as well as efficiency (delegated state update fails faster as it's usually local)
4. in case of any failure there will be no inconsistencies as state updates and snapshots are done by the same (Task) thread

I'll use the same order in all places.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogAggregatingState.java:
```diff
@@ -67,18 +73,24 @@ public OUT get() throws Exception {
     @Override
     public void add(IN value) throws Exception {
         delegatedState.add(value);
+        changeLogger.valueUpdated(delegatedState.getInternal(), getCurrentNamespace());
```
The same as `ChangelogReducingState`, update changelog first and then add to delegated state.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -51,16 +63,28 @@ public UV get(UK key) throws Exception {
+                    },
+                    getCurrentNamespace());
+        } else {
+            changeLogger.valueAdded(singletonMap(key, value), getCurrentNamespace());
```
I wonder is it possible that the value serializer of chenagelogMap state is not a `MapSerializer`?

rkhachatryan replied:
I see some checks in both Heap and RocksDB backends that disallow other serializers.
But it seems an implementation detail and I'm not sure that we can depend on it.
WDYT?

Myasuka replied:
In general, I don't think it's an implementation detail. `MapStateDescriptor` only allows `MapSerializer` or `MapTypeInfo`, and the only serializer would be created in `MapTypeInfo` is `MapSerializer`: see code [here](https://github.com/apache/flink/blob/4d73310ce6152e4d6bdccca44b8e1f8ecec0e1b2/flink-core/src/main/java/org/apache/flink/api/java/typeutils/MapTypeInfo.java#L115).

rkhachatryan replied:
I'm not sure (as `MapStateDescriptor` is not the only possible `StateDescriptor`); but it simplifies the code :) I'll remove the additional logic, thanks

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogListState.java:
```diff
@@ -39,53 +44,83 @@
+                                    .serialize(value, w),
+                    getCurrentNamespace());
+        } else {
+            changeLogger.valueAdded(singletonList(value), getCurrentNamespace());
```
I wonder is it possible that the value serializer of chenagelogList state is not a `ListSerializer`?

Myasuka replied:
The same as above.

Myasuka reviewed (commented):
Please take a look at my comments :)

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogListState.java:
```diff
@@ -39,53 +44,83 @@
+                    getCurrentNamespace());
+        } else {
+            changeLogger.stateAdded(singletonList(value), getCurrentNamespace());
+        }
```
Sync up offline.

I am confused with this part.

 if (getValueSerializer() instanceof ListSerializer    is not true,
adding singletonList will also be a problem?

That says when would be the case where getValueSerializer() is not ready
but `changeLogger.stateAdded(singletonList(value), getCurrentNamespace());` can be successful.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -96,12 +164,39 @@ public boolean isEmpty() throws Exception {
+        return (MapSerializer<UK, UV>) getValueSerializer();
+    }
+
+    private final BiConsumerWithException<
+                    Map.Entry<UK, UV>, DataOutputViewStreamWrapper, IOException>
+            changeWriter = (entry, out) -> serializeKey(entry.getKey(), out);
```
Move this to the top of this class...

Confusing to read.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java:
```diff
@@ -0,0 +1,122 @@
+            @Override
+            public UV setValue(UV value) {
+                try {
+                    changeLogger.valueElementChanged(out -> changeWriter.accept(entry, out), ns);
```
Do you need to log the value as well?

If you look at the how `changeWriter` is defined in ChangelogMapState, it only logs the key.

```
private final BiConsumerWithException<
                    Map.Entry<UK, UV>, DataOutputViewStreamWrapper, IOException>
            changeWriter = (entry, out) -> serializeKey(entry.getKey(), out);
```

rkhachatryan replied:
Right!

curcur commented:
Thanks @rkhachatryan for addressing the comments!

Some remaining points to discuss offline
1. why "MERGE_NS" is named "stateMerged" instead of "mergeNamespaces"
2. Where to loggingIterator/loggingIterable/loggingMapEntry should be put within the interface of "StateChangeLogger"
3. ChangelogAggregatingState.add()/ChangelogReducingState.add() using update instead of add to implement, might have differences on ChangelogAggregatingState. The reason is to log incremental changes instead of updated states. Details see in-line comment.
4. more explicit options.
MAP_PUT
LIST_APPEND
MAP_REMOVE
....
5. About Iterator Discussion
1). ChangelogKeyGroupedPriorityQueue.iterator(); is it changeable?
2). When an iterator can be changed. remove and update entry?
(Most of my concerns have already been addressed in last round of review, just double confirm)

6. How is update different from updateInternal
ChangelogListState.update/ChangelogListState.updateInternal
From reading their comments, it seems they are different somehow, but the changelogger seems to be same.

7. Specific question on map iterator. See in-line comments for details.
Some backend supports iterator change/remove; but some does not, but this is not the main point, I am asking in the case where iterator change/remove are supported 

In the case where the iterator supports changes
Let's say I get an iterator list of keys, and I get an iterator list of values, I remove some keys in the key iterator, and some values in the value iterator

Is it guaranteed that such remove is always paired? (Key, Value)

curcur commented:
For 7:

This is what I mean
MapState
    entries, keys, values; 
    keys = entries.getKeys();
    values = entries.getValues();
    
    keys.remove(some key)
    values.remove(some value)

Double check for rocksdb.

curcur commented:
For 6:

SET
SET_INTERNAL

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -0,0 +1,150 @@
+        /** Scope: key + namespace, also affecting other (source) namespaces. */
+        MERGE_NS((byte) 3),
+        /** Scope: key + namespace + element (e.g. user map key put or list append). */
+        CHANGE_ELEMENT((byte) 4),
+        /** Scope: key + namespace + element (e.g. user map remove or iterator remove). */
+        REMOVE_ELEMENT((byte) 5),
```
I agree. Replaced element-related operations with:
 - `ADD_ELEMENT`
 - `ADD_OR_UPDATE_ELEMENT`
 - `REMOVE_ELEMENT`
 - `REMOVE_FIRST_ELEMENT`

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogListState.java:
```diff
@@ -39,53 +44,83 @@
     public void updateInternal(List<V> valueToStore) throws Exception {
+        changeLogger.stateUpdated(valueToStore, getCurrentNamespace());
         delegatedState.updateInternal(valueToStore);
     }
```
TLDR: adding a new operation `SET_INTERNAL`

As discussed offline: `update` treats the provided value as user-supplied and so it performs some checks and protective copying (e.g. for heap state backend).

If left as is, then on recovery a different method might be called (`updateInternal` instead of `update` or vice versa). It probably won't cause any issues for now, but I'll separate these operations to be on the safe side.

rkhachatryan commented:
Thanks for the thorough reviews @curcur and @Myasuka !
I addressed most of the comments, PTAL.

Regarding the state mutation via iterators, @curcur :
- both RocksDB and Heap allow to update the state via iterators
- multiple updates via different iterators may or may not fail; but if they succeed then the changes should be reflected by logging iterators in the correct order (assuming single-threaded environment)

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -0,0 +1,172 @@
+    private byte[] serializeRaw(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter)
+            throws IOException {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
```
I believe you will add TODO for performance in https://github.com/apache/flink/pull/15420, right?

rkhachatryan replied:
Right (in 98d8eadd9fa4a52e6f8eaa7b62458a9d9cf76bc8).

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -0,0 +1,172 @@
+                wrapper -> {
+                    wrapper.writeByte(op.code);
+                    serializeScope(ns, wrapper);
+                    dataWriter.accept(wrapper);
```
Since some `dataWriter` could be empty as `out -> {}`, I suppose it might help improve a bit performance by passing a `null` dataWriter here and check whether this is null to avoid creating a new lambda `out -> {}` here.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +108,72 @@ public boolean contains(UK key) throws Exception {
-        return delegatedState.keys();
+        Iterable<UK> iterable = delegatedState.keys();
+        return iterable == null
+                ? null
```
First of all, the `delegatedState` would not return `null`. Secondly, if we want a check here, we should better return `Collections::emptyIterator` here.

rkhachatryan replied:
Hmm..I was sure that `null` is an allowed return value here, but you are right.
I'll just remove the null check.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLoggingIterator.java:
```diff
@@ -0,0 +1,81 @@
+                    removalWriter,
+            Namespace ns) {
+        return iterator == null
+                ? null
```
We should return `Collections.emptyIterator()` here and tag this method as `@Nonnull` which aligns with `InternalPriorityQueue#iterator`

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/KvStateChangeLogger.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.IOException;
+import java.util.Collection;
+
+interface KvStateChangeLogger<State, Namespace> extends StateChangeLogger<State, Namespace> {
```
I think a javadoc telling that this changelogger is not only used for value state but all kinds of keyed state is better. Or maybe another name which would let developer mix with the usage of value state.

rkhachatryan replied:
Good point, I'll add javadoc.

Myasuka reviewed (commented):
overall, I think this PR looks good to me. The main comment is that we should return an empty iterator instead of `null`.

rkhachatryan commented:
I addressed the remaining issues @Myasuka, PTAL (I've removed the `null` check when creating an iterator).

Myasuka reviewed (approved):
Glad to see all comments addressed, LGTM.

curcur commented:
I am fine as long as all the comments are addressed. Take a glance. approved.

curcur commented:
Thanks @rkhachatryan !
